### PR TITLE
feat: Infer ScheduledEvent.entity_type

### PR DESCRIPTION
**Pull request by zachnieto**
## Summary

Infer ScheduledEvent.entity_type from provided Channel type when creating ScheduledEvent #590

## Checklist

- [X] If code changes were made, then they have been tested
    - [X] I have updated the documentation to reflect the changes
    - [X] I have formatted the code properly by running `task lint`
    - [X] I have type-checked the code by running `task pyright`
- [ ] This PR fixes an issue
- [X] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: disnake/guild.py [hunk 1]
```diff
@@ -2224,14 +2232,26 @@ async def create_scheduled_event(
...
+
+        if channel is None:
+            entity_type = GuildScheduledEventEntityType.external
+        elif hasattr(channel, "type") and isinstance(channel.type, ChannelType):  # type: ignore
+            if channel.type == ChannelType.voice:  # type: ignore
+                entity_type = GuildScheduledEventEntityType.voice
+            elif channel.type == ChannelType.stage_voice:  # type: ignore
+                entity_type = GuildScheduledEventEntityType.stage_instance
+        elif entity_type is MISSING:
+            raise TypeError("entity_type must be provided if the channel does not have a type")
```
Comment (onerandomusername) [inline disnake/guild.py:2253]: This logic has a bit of an issue where the channel has a type that isn't stage_voice or voice. 

We should also raise TypeError if the type isn't voice or stage_voice.

File: disnake/guild.py [hunk 2]
```diff
@@ -2224,14 +2232,26 @@ async def create_scheduled_event(
...
+            could not be assumed from the ``channel``.
 
         Returns
         -------
         :class:`GuildScheduledEvent`
             The newly created guild scheduled event.
         """
+
+        if channel is None:
+            entity_type = GuildScheduledEventEntityType.external
```
Comment (onerandomusername) [inline disnake/guild.py:2246]: if we set `channel` to MISSING here we don't need the second None check later.

File: disnake/guild.py [hunk 3]
```diff
@@ -3427,7 +3447,6 @@ async def edit_role_positions(
 
         role_positions: List[Any] = []
         for role, position in positions.items():
-
```
Comment (onerandomusername) [inline disnake/guild.py:3430]: this change seems unrelated?

**Review by onerandomusername — CHANGES_REQUESTED** [review]
This pretty good. I found two bugs while testing, but they should be easy fixes. I also left a lengthy note about the changelog notes 😅 which is partially unique to this situation.

File: disnake/guild.py [hunk 4]
```diff
@@ -2163,6 +2163,13 @@ async def create_scheduled_event(
 
         Creates a :class:`GuildScheduledEvent`.
 
+        If ``channel`` is provided and has a type matching to a type of :class:`GuildScheduledEventEntityType`,
```
Comment (shiftinv) [inline disnake/guild.py:2166]: This might be better represented using a table:
(once again can't directly add a suggestion)
```diff
diff --git a/disnake/guild.py b/disnake/guild.py
index 795e7553..60579ca3 100644
--- a/disnake/guild.py
+++ b/disnake/guild.py
@@ -2163,18 +2163,17 @@ class Guild(Hashable):
 
         Creates a :class:`GuildScheduledEvent`.
 
-        If ``channel`` is provided and has a type matching to a type of :class:`GuildScheduledEventEntityType`,
-            the ``entity_type`` is not required and will automatically match.
+        Based on the channel/entity type, there are different restrictions regarding
+        other parameter values, as shown in this table:
 
-        If ``channel`` is not provided, then ``entity_type`` must be provided.
-
-        If ``channel`` is None, then ``entity_type`` is assumed to be :class:`GuildScheduledEventEntityType.external`
+        .. csv-table::
+            :header: "``channel``", "``entity_type``", "``scheduled_end_time``", "``entity_metadata`` with location"
 
-        If ``entity_type`` is :class:`GuildScheduledEventEntityType.external`:
+            :class:`VoiceChannel`, automatically set to :attr:`~GuildScheduledEventEntityType.voice`,          optional, unset
+            :class:`StageChannel`, automatically set to :attr:`~GuildScheduledEventEntityType.stage_instance`, optional, unset
+            unspecified snowflake, required,                                                                   optional, unset
+            ``None`` or unset,     automatically set to :attr:`~GuildScheduledEventEntityType.external`,       required, required
 
-        - ``channel`` should not be set
-        - ``entity_metadata`` with a location field must be provided
-        - ``scheduled_end_time`` must be provided
 
         .. versionadded:: 2.3
 

```

I'm a bit unsure about the "unspecified snowflake" phrasing though, perhaps someone else has a better idea for that.
In any case, looks like this:
image

File: disnake/guild.py [hunk 5]
```diff
@@ -2224,14 +2232,29 @@ async def create_scheduled_event(
...
 
         Returns
         -------
         :class:`GuildScheduledEvent`
             The newly created guild scheduled event.
         """
+
+        if channel is None:
+            entity_type = GuildScheduledEventEntityType.external
+            channel = MISSING
```
Comment (shiftinv) [inline disnake/guild.py:2247]:
```suggestion
        if channel is None or channel is MISSING:
            entity_type = GuildScheduledEventEntityType.external
```
(combined with the other channel change below)

File: disnake/guild.py [hunk 6]
```diff
@@ -2262,7 +2285,7 @@ async def create_scheduled_event(
             fields["image"] = await utils._assetbytes_to_base64_data(image)
 
         if channel is not MISSING:
-            fields["channel_id"] = channel.id
+            fields["channel_id"] = channel.id  # type: ignore
```
Comment (shiftinv) [inline disnake/guild.py:2288]:
```suggestion
        if channel:
            fields["channel_id"] = channel.id
```

File: disnake/guild.py [hunk 7]
```diff
@@ -2224,14 +2232,29 @@ async def create_scheduled_event(
...
         """
+
+        if channel is None:
+            entity_type = GuildScheduledEventEntityType.external
+            channel = MISSING
+        elif hasattr(channel, "type") and isinstance(channel.type, ChannelType):  # type: ignore
+            if channel.type == ChannelType.voice:  # type: ignore
+                entity_type = GuildScheduledEventEntityType.voice
+            elif channel.type == ChannelType.stage_voice:  # type: ignore
+                entity_type = GuildScheduledEventEntityType.stage_instance
```
Comment (shiftinv) [inline disnake/guild.py:2252]:
```suggestion
        elif isinstance(channel_type := getattr(channel, "type", None), ChannelType):
            if channel_type == ChannelType.voice:
                entity_type = GuildScheduledEventEntityType.voice
            elif channel_type == ChannelType.stage_voice:
                entity_type = GuildScheduledEventEntityType.stage_instance
```
To avoid the `# type: ignore`s :p

**Review by shiftinv — COMMENTED** [review]
Thanks for taking this on, already looking pretty good!
It might be a good idea to change `GuildScheduledEvent.edit` in a similar way, though we should finalize the general concept first before copying it over to that :thinking:

File: disnake/guild.py (hunk 1 above)
Comment (onerandomusername) [reply disnake/guild.py]: fixed in https://github.com/DisnakeDev/disnake/pull/617/commits/119f3081fc04dbe12c52401d83cf8042f4c12eb5

File: disnake/guild.py (hunk 5 above)
Comment (onerandomusername) [reply disnake/guild.py:2247]: IMHO this shouldn't work like this. If channel is unprovided, I think we should require that entity_type was provided rather than default to external.

Comment (shiftinv) [reply disnake/guild.py:2247]: It'll require an additional overload, but in the long run that behavior is better, yeah. I've updated my other comments to reflect this and not default to external if no channel is given.

File: disnake/guild.py [hunk 8]
```diff
@@ -2131,12 +2131,12 @@ async def create_scheduled_event(
         self,
```
[the remark this answers is not in the document]
Comment (zachnieto) [reply disnake/guild.py:2131]: Fixed in [fce1fd9](https://github.com/DisnakeDev/disnake/pull/617/commits/fce1fd99f4eb82473fad14e3ce3a417d5fd88e4e)

File: disnake/guild.py [hunk 9]
```diff
@@ -2131,12 +2131,12 @@ async def create_scheduled_event(
...
         *,
         name: str,
+        channel: Optional[Snowflake],
```
[the remark this answers is not in the document]
Comment (zachnieto) [reply disnake/guild.py]: fixed in [fce1fd9](https://github.com/DisnakeDev/disnake/pull/617/commits/fce1fd99f4eb82473fad14e3ce3a417d5fd88e4e)

File: disnake/guild.py (hunk 4 above)
Comment (zachnieto) [reply disnake/guild.py]: changed in [fce1fd9](https://github.com/DisnakeDev/disnake/pull/617/commits/fce1fd99f4eb82473fad14e3ce3a417d5fd88e4e)

File: disnake/guild.py (hunk 5 above)
Comment (zachnieto) [reply disnake/guild.py:2269]: fixed in [fce1fd9](https://github.com/DisnakeDev/disnake/pull/617/commits/fce1fd99f4eb82473fad14e3ce3a417d5fd88e4e)

**Review by onerandomusername — COMMENTED** [review]
This looks good.

Shiftinv mentioned the idea of adding these changes to GuildScheduledEvent.edit(), would you like to make those in this pull as well, or would you rather this is a second pr? Either is fine, you can even open the second pr if you'd like.

**Comment by zachnieto** [comment]
> This looks good.
> 
> Shiftinv mentioned the idea of adding these changes to GuildScheduledEvent.edit(), would you like to make those in this pull as well, or would you rather this is a second pr? Either is fine, you can even open the second pr if you'd like.

Might as well do it all here !

**Comment by zachnieto** [comment]
First time doing overloads in python, hopefully im getting close lol

File: disnake/guild_scheduled_event.py [hunk 10]
```diff
@@ -373,6 +391,12 @@ async def edit(
         - ``entity_metadata`` with a location field must be provided
         - ``scheduled_end_time`` must be provided
 
+        If updating ``channel`` to ``None``, ``entity_type`` will be set to
+        :attr:`GuildScheduledEventEntityType.external`.
+
+        If updating ``channel`` to :class:`.abc.Snowflake`, ``entity_type`` will be set to the channel's type
+        if the type is either :class:`ChannelType.voice` or :class:`ChannelType.stage_voice`.
```
Comment (onerandomusername) [inline disnake/guild_scheduled_event.py:398]: missing a versionchanged, also should update the cases when TypeError is raised.

File: disnake/guild_scheduled_event.py [hunk 11]
```diff
@@ -48,6 +48,7 @@
 if TYPE_CHECKING:
     from .abc import GuildChannel, Snowflake
     from .asset import AssetBytes
+    from .enums import ChannelType
```
Comment (onerandomusername) [inline disnake/guild_scheduled_event.py:51]: This is required at runtime and should be imported outside of this block.

**Review by onerandomusername — CHANGES_REQUESTED** [review]
A couple comments, one import error, but after fixing the code seems good, leaving everything to the documentation.

**Review by onerandomusername — APPROVED** [review]
lgtm

File: disnake/guild_scheduled_event.py [hunk 12]
```diff
@@ -348,6 +349,23 @@ async def edit(
...
+        name: str = ...,
+        description: Optional[str] = ...,
+        image: Optional[AssetBytes] = ...,
+        privacy_level: GuildScheduledEventPrivacyLevel = ...,
+        scheduled_start_time: datetime = ...,
+        scheduled_end_time: Optional[datetime] = ...,
+        status: GuildScheduledEventStatus = ...,
+        reason: Optional[str] = ...,
+    ) -> GuildScheduledEvent:
+        ...
```
Comment (shiftinv) [inline disnake/guild_scheduled_event.py:367]: To cover all cases, I believe we need an additional overload here - essentially, the end goal for these overloads is:
1. generic, no type change (should accept all parameters)
2. `entity_type=external` specified, `channel=None` may be provided but doesn't have to be
3. `entity_type=voice/stage` specified, `channel=<valid>` may be provided but doesn't have to be
4. `channel=None` specified
5. `channel=<valid>` specified

(I hope that makes sense, and that I didn't miss anything)

A `channel: None = ...,` is also needed in overload no.2 (can't suggest it directly because GitHub) to match the above description and allow calls like
```py
await e.edit(
    entity_type=disnake.GuildScheduledEventEntityType.external,
    channel=None,
    entity_metadata=disnake.GuildScheduledEventMetadata(location="h"),
)
```

These would be (4) and (5) in the list above:
```suggestion
    # channel=None, no entity_type
    @overload
    async def edit(
        self,
        *,
        channel: None,
        name: str = ...,
        description: Optional[str] = ...,
        image: Optional[AssetBytes] = ...,
        privacy_level: GuildScheduledEventPrivacyLevel = ...,
        scheduled_start_time: datetime = ...,
        scheduled_end_time: datetime = ...,
        entity_metadata: GuildScheduledEventMetadata = ...,
        status: GuildScheduledEventStatus = ...,
        reason: Optional[str] = ...,
    ) -> GuildScheduledEvent:
        ...

    # valid channel, no entity_type
    @overload
    async def edit(
        self,
        *,
        channel: Snowflake,
        name: str = ...,
        description: Optional[str] = ...,
        image: Optional[AssetBytes] = ...,
        privacy_level: GuildScheduledEventPrivacyLevel = ...,
        scheduled_start_time: datetime = ...,
        scheduled_end_time: Optional[datetime] = ...,
        status: GuildScheduledEventStatus = ...,
        reason: Optional[str] = ...,
    ) -> GuildScheduledEvent:
        ...
```

File: disnake/guild_scheduled_event.py [hunk 13]
```diff
@@ -441,6 +466,18 @@ async def edit(
...
+            entity_type = GuildScheduledEventEntityType.external
+        elif channel is not MISSING and isinstance(
+            channel_type := getattr(channel, "type", None), ChannelType
+        ):
+            if channel_type is ChannelType.voice:
+                entity_type = GuildScheduledEventEntityType.voice
+            elif channel_type is ChannelType.stage_voice:
+                entity_type = GuildScheduledEventEntityType.stage_instance
+            else:
+                raise TypeError("channel type must be either 'stage' or 'stage_voice'")
```
Comment (shiftinv) [inline disnake/guild_scheduled_event.py:479]:
```suggestion
            else:
                raise TypeError("channel type must be either 'voice' or 'stage_voice'")
```

File: changelog/590.breaking.rst [hunk 14]
```diff
@@ -0,0 +1 @@
+548.breaking.rst
```
Comment (shiftinv) [inline changelog/590.breaking.rst:1]: I'm not really seeing how this change is breaking - on the surface it essentially just removes the requirement of `entity_type` if a corresponding `channel` is given instead, doesn't it?
cc @onerandomusername

File: changelog/548.breaking.rst [hunk 15]
```diff
@@ -1 +1,2 @@
 Rename ``channel_id`` parameter to ``channel`` on :attr:`Guild.create_scheduled_event` and :meth:`GuildScheduledEvent.edit`.
+Allow ``entity_type`` parameter :attr:`Guild.create_scheduled_event` to be missing.
```
Comment (shiftinv) [inline changelog/548.breaking.rst:2]: Fairly sure this line would be better suited in a `590.feature.rst`.

**Review by shiftinv — COMMENTED** [review]
A few more comments - I hope it all makes sense, writing overloads for these type-dependent fields is a bit complex but I think that should be it now :sweat_smile:

File: changelog/590.breaking.rst (hunk 14 above)
Comment (onerandomusername) [reply changelog/590.breaking.rst:1]: This is a symlink to 548's change note, which was breaking, and these changes were added into that changefragment. See [my comment above](https://github.com/DisnakeDev/disnake/pull/617#discussion_r924086077) as to why it is like this.

File: changelog/548.breaking.rst (hunk 15 above)
Comment (onerandomusername) [reply changelog/548.breaking.rst:2]: If we do move this, this pr should still continue to symlink 548.breaking.rst

Comment (zachnieto) [reply changelog/548.breaking.rst:2]: so we mean to remove the symlink and just move the changelog entry to `590.feature.rst`?

Comment (onerandomusername) [reply changelog/548.breaking.rst:2]: Keep the symlink, but move this snippet to `590.feature.rst`

Comment (onerandomusername) [reply changelog/548.breaking.rst]: https://github.com/DisnakeDev/disnake/pull/617/commits/1ee428f45e2ab263731f5d9b38e8b4f562930059

File: disnake/guild_scheduled_event.py [hunk 16]
```diff
@@ -373,6 +391,12 @@ async def edit(
         - ``entity_metadata`` with a location field must be provided
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply disnake/guild_scheduled_event.py]: https://github.com/DisnakeDev/disnake/pull/617/commits/b9f9e9a247b56b4697e0adcccfd134da00cc20b3

File: disnake/guild_scheduled_event.py (hunk 10 above)
Comment (shiftinv) [reply disnake/guild_scheduled_event.py]: Looks like the `versionchanged` sections here and on `create_scheduled_event` got lost in a commit somewhere

**Review by shiftinv — COMMENTED** [review]
(this should be it, hope I didn't miss anything this time around)

**Review by searchify7 — APPROVED** [review]

File: disnake/guild.py [hunk 17]
```diff
@@ -2163,11 +2180,19 @@ async def create_scheduled_event(
...
 
-        - ``channel`` should not be set
-        - ``entity_metadata`` with a location field must be provided
-        - ``scheduled_end_time`` must be provided
+        .. csv-table::
+            :widths: 25, 30, 20, 25
+            :header: "``channel``", "``entity_type``", "``scheduled_end_time``", "``entity_metadata`` with location"
+
+            :class:`VoiceChannel`, :attr:`~GuildScheduledEventEntityType.voice` (set automatically), optional, unset
+            :class:`StageChannel`, :attr:`~GuildScheduledEventEntityType.stage_instance` (set automatically), optional, unset
```
Comment (onerandomusername) [inline disnake/guild.py:2192]: these are slightly inaccurate, and should refer to the type attribute on the object instead. Passing a PartialMessageable with the voice channel type would work, for example. Passing an `Object` with the type attribute set to `ChannelType.voice` would also work. The class doesn't actually matter, and this distinction is important.

**Review by onerandomusername — CHANGES_REQUESTED** [review]
tiny comment, bit about the table wording

File: disnake/guild.py (hunk 17 above)
Comment (zachnieto) [reply disnake/guild.py]: I can see it being a bit confusing to users, hopefully my wording isnt too rough

**Review by shiftinv — APPROVED** [review]
lgtm!

File: disnake/guild.py (hunk 17 above)
Comment (onerandomusername) [reply disnake/guild.py]: I don't think so.

**Review by onerandomusername — APPROVED** [review]
thanks for this!
